### ISSUE TITLE
Feedback: Java RMI 2016

### ISSUE THREAD

**Issue by rolandweber** [rolandweber/pityoulish#48]
tl;dr: The session ran completely out of time. I have to prepare totally different next year. See further below for my ideas.

I started by explaining the steps for development in general: define API, generate stubs and skeletons, export objects, use the registry. Over 30 minutes of dry talk, and by the time the students got to that point in the programming session, they probably didn't remember a thing. _Doesn't work._

I let the programming session run close to the end of the lesson, almost 120 minutes. After 45 minutes, the first group started on the second part, while others were still struggling with locating the registry. After 60 minutes, about half of the students were working on the second part. Nobody completed both parts, but nobody wanted to continue either. One student asked about the server later, to complete the exercise at home. During the last minutes, I gathered feedback:
* Some hands-on programming is useful, compared to just listening. (1)
* It didn't feel like applying the things I had explained about remote calls. (2)
_I agree. The theory was about stubs, skeletons, and sending messages. But all that happens under the cover. The programming was about looking up objects in the registry, and exporting one._
* It didn't feel like programming at all, just debugging. (2)
_Yes, that's how the sessions are designed. In real life, a lot of programming is debugging, too. But I thought the students would get the big picture while looking at the pieces. That's not the case._
* How about programming an application from scratch? (1)
_Takes too long for the classroom. Could do it as homework, but that would have to be voluntary. The programming is not relevant for the grade, and I don't want the minimum-effort students to spend their preparation time on programming instead of learning the relevant material. Actually, I don't want to spend my preparation time on a completely optional programming exercise either._

During the programming session, two students asked about the meaning of the marker. Don't know how many others were wondering without asking. _I thought that the functionality of the Message Board would be familiar by now, from the sockets exercise. Not so._

One student cloned the GitHub repo with the full source, instead of downloading the JARs with the processed sources. _Should I set up temporary GitHub repos with the processed source for the exercises?_ #50

Several students were struggling with `RegistryBackendHandlerImpl.getTicketIssuer()`, without noticing `getMessageBoard()` just a few lines above - although it was right there on their screen. Not sure if any of them looked at the JavaDocs of the interface and noticed the similar purpose of these methods. _No idea how to overcome that._

Not everybody was familiar with typecasts. _At least one IDE showed an auto-correct option to add the typecast. But it's not the idea of the exercise to select from auto-correct options._

The programs should print a success message for operations that have no other output, like "return ticket" and "put message". #49

There were problems with desktop firewalls on corporate laptops. The firewall blocked the remote calls from the server to the outlet, so it wasn't possible to publish the outlet. The program just hung for minutes, apparently the server didn't even get an error when trying to connect. For some, but not all, it worked when exorting the outlet on port 21.

# For 2017...

I must present the functionality of the Message Board in class. Week 1, or at latest before the sockets exercise. Students have too much other information to digest during the programming sessions. I have to give them the big picture before (or after).

I should approach the RPC/RMI topic from a completely different angle. Instead of theory, start with the programming exercise. Guide the students by switching between explanations and programming. In exercise order, rather than chronologically:
* There's an object on the server, it's in the registry, look it up and make a call. -> _programming_
* Export your own object and publish it on the server. -> _programming_
* Call other students' objects, handle errors. -> _programming_
* Clean up your object to exit cleanly. -> _programming_

Don't make the programming/debugging blocks too small. Students should have enough time to dive into the code, they shouldn't just wait until the solution is presented.

Explain background and theory of stubs, skeletons, generating both, serializing data,... afterwards. Maybe even the week after.

**Comment by rolandweber** [comment]
A follow-the-board client would be useful during the exercise, see #51.
The RMI server does not print message content (and the sockets server probably shouldn't either).

**Comment by rolandweber** [comment]
A key part of RPCs is that stubs and skeletons are automatically generated from the API description. Java RMI with Reflections has made this step completely optional. That might be an argument for using gRPC in class, see #36.

At least, the Message Board with Java RMI has some screen output where the term "Proxy" appears.

In previous years, I live-coded an unrelated example on the projector. But that left students completely baffled, as they were only watching and not coding themselves.
